File: kdb-core/src/main/java/com/t13max/kdb/serial/SerialTask.java
package com.t13max.kdb.serial;

import com.t13max.kdb.consts.Const;
import lombok.Getter;

/**
 * 串行任务
 *
 * @author t13max
 * @since 11:24 2025/7/10
 */
@Getter
public abstract class SerialTask implements Runnable {

    //唯一id
    protected final long id;

    //是否存在IO操作
    protected final boolean io;

    //作用域
    protected final String scope;

    protected SerialTask(long id) {
        this(id, false, Const.SCOPE_ROLE);
    }

    protected SerialTask(long id, boolean io) {
        this(id, io, Const.SCOPE_ROLE);
    }

    protected SerialTask(long id, boolean io, String scope) {
        this.id = id;
        this.io = io;
        this.scope = scope;
    }

    @Override
    public final void run() {
        try {
            if (process()) {
                commit();
                return;
            }
            rollback();
        } catch (Throwable throwable) {
            rollback();
        } finally {
            finished();
        }
    }

    private void finished() {

    }

    private void rollback() {

    }

    private void commit() {

    }

    protected abstract boolean process();
}

File: kdb-core/src/main/kotlin/com/t13max/kdb/serial/SerialTask.kt
package com.t13max.kdb.serial

import com.t13max.kdb.consts.Const

/**
 * 串行任务
 *
 * @author t13max
 * @since 14:15 2025/8/16
 */
abstract class SerialTask(
    val id: Long,
    val io: Boolean = false,
    val scope: String = Const.SCOPE_ROLE
) : Runnable {

    constructor(id: Long) : this(id, false, Const.SCOPE_ROLE)

    constructor(id: Long, io: Boolean) : this(id, io, Const.SCOPE_ROLE)

    final override fun run() {
        try {
            if (process()) {
                commit()
                return
            }
            rollback()
        } catch (t: Throwable) {
            rollback()
        } finally {
            finished()
        }
    }

    private fun finished() {

    }

    private fun rollback() {

    }

    private fun commit() {

    }

    protected abstract fun process(): Boolean
}